Mathlib/Dynamics/TopologicalEntropy/DynamicalEntourage.lean

```diff
@@ -0,0 +1,132 @@
+/-
+Copyright (c) 2024 Damien Thomine. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damien Thomine, Pietro Monticone
+-/
+import Mathlib.Order.Interval.Finset.Nat
+import Mathlib.Order.OmegaCompletePartialOrder
+import Mathlib.Topology.UniformSpace.Basic
+
+/-!
+# Dynamical entourages
+Bowen-Dinaburg's definition of topological entropy of a transformation `T` in a metric space
+`(X, d)` relies on the so-called dynamical balls. These balls are sets
+`B (x, ε, n) = { y | ∀ k < n, d(T^[k] x, T^[k] y) < ε }`.
+
+We implement Bowen-Dinaburg's definitions in the more general context of uniform spaces. Dynamical
+balls are replaced by what we call dynamical entourages. This file collects all general lemmas
+about these objects.
+
+## Main definitions
+- `dynEntourage`: dynamical entourage associated with a given transformation `T`, uniformity `U`
+and time `n`.
+
+## Tags
+entropy
+
+## TODO
+Once #PR14718 has passed, add product of uniformities.
+
+In the context of (pseudo-e)metric spaces, relate the usual definition of dynamical balls with
+these dynamical uniformities.
+-/
+
+namespace DynamicalEntourage
+
+open Prod Set Uniformity UniformSpace
+
+variable {X : Type*}
+
+/-- The dynamical entourage associated to a transformation `T`, enourage `U` and time `n` is the set
+of points `(x, y)` such that `(T^[k] x, T^[k] y) ∈ U` for all `k < n`, i.e. which are `U`-close up
+to time `n`.-/
+def dynEntourage (T : X → X) (U : Set (X × X)) (n : ℕ) : Set (X × X) :=
+  ⋂ k < n, (map T T)^[k] ⁻¹' U
+
+lemma dynEntourage_eq_inter_Ico (T : X → X) (U : Set (X × X)) (n : ℕ) :
+    dynEntourage T U n = ⋂ k : Ico 0 n, (map T T)^[k] ⁻¹' U := by
+  simp [dynEntourage, iInter_coe_set, mem_Ico, zero_le, true_and]
+
+lemma mem_dynEntourage {T : X → X} {U : Set (X × X)} {n : ℕ} {x y : X} :
+    (x, y) ∈ dynEntourage T U n ↔ ∀ k < n, (T^[k] x, T^[k] y) ∈ U := by
+  simp [dynEntourage, map_iterate, mem_preimage, mem_iInter, map_apply]
+
+lemma mem_ball_dynEntourage {T : X → X} {U : Set (X × X)} {n : ℕ} {x y : X} :
+    y ∈ ball x (dynEntourage T U n) ↔ ∀ k < n, T^[k] y ∈ ball (T^[k] x) U := by
+  simp [ball, mem_preimage]; exact mem_dynEntourage
+
+lemma dynEntourage_mem_uniformity [UniformSpace X] {T : X → X} (h : UniformContinuous T)
+    {U : Set (X × X)} (U_uni : U ∈ 𝓤 X) (n : ℕ) :
+    dynEntourage T U n ∈ 𝓤 X := by
+  rw [dynEntourage_eq_inter_Ico T U n]
+  refine Filter.iInter_mem.2 fun k ↦ ?_
+  rw [map_iterate T T k]
+  exact uniformContinuous_def.1 (UniformContinuous.iterate T k h) U U_uni
+
+lemma idRel_subset_dynEntourage (T : X → X) {U : Set (X × X)} (h : idRel ⊆ U) (n : ℕ) :
+    idRel ⊆ (dynEntourage T U n) := by
+  simp only [dynEntourage, map_iterate, subset_iInter_iff, idRel_subset, mem_preimage, map_apply]
+  exact fun _ _ _ ↦ h rfl
+
+lemma SymmetricRel.dynEntourage (T : X → X) {U : Set (X × X)} (h : SymmetricRel U) (n : ℕ) :
```

Comment:
```suggestion
lemma _root_.SymmetricRel.dynEntourage (T : X → X) {U : Set (X × X)} (h : SymmetricRel U) (n : ℕ) :
```
Since you're in a namespace, you need `_root_` to get out of this namespace.